nixos/modules/services/networking/crab-hole.nix

```
@@ -0,0 +1,230 @@
+                      ];
+                      example = "udp";
+                      default = "udp";
+                      description = "The protocol to listen on";
+                    };
+                    listen = lib.mkOption {
+                      type = lib.types.str;
+                      example = "[::]";
+                      default = "localhost";
+                      description = "The IP to listen on";
+                    };
+                    port = lib.mkOption {
+                      type = lib.types.ints.u16;
+                      example = 53;
+                      default = 53;
+                      description = "The port to listen on";
+                    };
+                  };
+                }
+              );
+            };
+
+            upstream = {
+              options = lib.mkOption {
+                default = { };
+                description = "Set resolover options. For a full list see https://docs.rs/trust-dns-resolver/0.23.0/trust_dns_resolver/config/struct.ResolverOpts.html";
+                type = lib.types.submodule {
+                  freeformType = settingsFormat.type;
+                };
+              };
+              name_servers = lib.mkOption {
+                description = "Configuration for the NameServers";
+                type = lib.types.listOf (
+                  lib.types.submodule {
+                    options = {
+                      protocol = lib.mkOption {
+                        type = lib.types.enum [
+                          "udp"
+                          "tls"
+                          "https"
+                          "quic"
+                        ];
+                        example = "tls";
+                        default = "tls";
+                        description = "The protocol to use when communicating with the NameServer";
+                      };
+                      socket_addr = lib.mkOption {
+                        type = lib.types.str;
+                        example = "1.1.1.1:853";
+                        description = "Address of the upstream DNS server";
+                      };
+                      tls_dns_name = lib.mkOption {
+                        type = lib.types.str;
+                        example = "1dot1dot1dot1.cloudflare-dns.com";
+                        description = "Domain name of the upstream DNS server";
+                      };
+                      trust_nx_responses = lib.mkEnableOption "Whether to trust `NXDOMAIN` responses from upstream nameservers";
+                    };
+                  }
+                );
+              };
+            };
+          };
+        };
+      };
+
+      configFile = lib.mkOption {
+        type = lib.types.path;
+        description = "The config file of crab-hole. If files are added via url, make sure the service has access to them. Setting this option will override any configuration applied by the settings option.";
```

Learn to love multiline strings
```suggestion
        description = ''
        	The config file of crab-hole. 
        	
        	If files are added via URLs, make sure the service has access to them. 
        	Setting this option will override any configuration applied by the settings option.
        '';
```